Implement FR #2338 for current branches / automatically advancing branches

emesterhazy opened the pull request:
This is an *experimental* feature that allows users to configure branches which will automatically advance when a new commit is created with `jj commit` or `jj new`. This feature may change or be removed in the future.

## Setup
This behavior can be enabled by default for all branches by setting
the following in the `config.toml`:

```
[experimental-advance-branches]
enabled-branches = ["glob:*"]
```

Specific branches can also be disabled:
```
[experimental-advance-branches]
enabled-branches = ["glob:*"]
disabled-branches = ["main", "glob:push-*"]
```

Branches that match a disabled pattern will not be advanced, even if they also
match an enabled pattern.

## How it works

- When `jj commit` and `jj new` are run, they target a specific revision, the "target revision". For `jj commit` this is always `@`, and for `jj new` the user specifies the target.
- When `jj commit` or `jj new` is run, jj will look at the parents of the target revision and advance any eligible branches pointing to them to the target revision (not the new commit). Branches are eligible to advance if the feature is enabled globally and the branch is not in the overrides list, or vice versa.
- Eligible branches will be advanced and will point to `@-` after `jj commit` or `jj new` completes.
- `jj new` only advances branches if the target is a single revision and the `--after` and `--before` are not used.

The result is that the branches move forward automatically as new commits are added. This is useful if you are committing linearly and don't want to move the branch pointer manually. Users familiar with Git may find that this feature mimics Git's behavior where branches move automatically as commits are added.

## Misc
An earlier version of this PR also included a feature for colocated Git repos:

> This feature also includes a convenience for colocated git repos. When the working commit is changed in a colocated repo and there is an advance-branches eligible branch pointing to its first parent, the Git HEAD is set to the branch ref instead of being detached at the parent commit id. This applies any time the working commit is changed, not just when `jj commit` is run.

I decided to split that feature into a separate PR, which I will mail after this PR is submitted.

emesterhazy commented:
@martinvonz @yuja Would you mind taking a look at this and sharing your thoughts when you get a chance?

We've been discussing variations of what I think is the same idea in https://github.com/martinvonz/jj/issues/2338. I think this implementation will work well for Git users, and still makes sense for jj users where the working commit (i.e. what's created by `jj commit`)is typically WIP analogous to an uncommitted working copy in Git.

martinvonz commented:
> I think this implementation will work well for Git users, and still makes sense for jj users where the working commit (i.e. what's created by `jj commit`)is typically WIP analogous to an uncommitted working copy in Git.

Thanks for mentioning that. I think it's important that we design something that works well in repos that are not colocated with Git. Once we have come up with something we like, we should try to make our code for colocated repos adapt to that model. So I think we should try to ignore how Git behaves when we design this feature, unless we say that this is a feature designed only for Git interop (so it would apply only when using the Git backend?).

> @martinvonz @yuja Would you mind taking a look at this and sharing your thoughts when you get a chance?

I think I'm back to preferring the model where we consider branches pointing to `@` as active, even if that means more work when exporting to Git.

emesterhazy commented:
> I think I'm back to preferring the model where we consider branches pointing to @ as active, even if that means more work when exporting to Git.

I think this means that when `jj commit` is run, only the branch pointing at `@` is eligible to advance, and it will be moved to the new commit created by `jj commit` is that right? The current PR does this for `@-` instead of `@`.

Can you elaborate on why you think it's better to use `@` instead of `@-`? I actually think that using `@-` as the target makes more sense because the new commit created by `jj commit` or `jj new` is usually either left empty (if you're done and pushing your work to Github or some other forge) or is an undescribed work-in-progress commit. I actually don't think I'd want the branch to move to the new `@` commit until I'm ready done with it and ready to `jj new` on top of it.

martinvonz commented:
> I think this means that when `jj commit` is run, only the branch pointing at `@` is eligible to advance, and it will be moved to the new commit created by `jj commit` is that right? The current PR does this for `@-` instead of `@`.

Correct.

> Can you elaborate on why you think it's better to use `@` instead of `@-`? I actually think that using `@-` as the target makes more sense because the new commit created by `jj commit` or `jj new` is usually either left empty (if you're done and pushing your work to Github or some other forge) or is an undescribed work-in-progress commit. I actually don't think I'd want the branch to move to the new `@` commit until I'm ready done with it and ready to `jj new` on top of it.

You might be right. I never use branches so I'm not sure what behavior people want. We generally don't treat `@-` differently. Maybe that's all that bothers me about this proposal. But again, I'm not the target audience for this.

yuja commented:
Just dropping random comments. I'm also not the target audience as I always start git with `checkout --detach`.

(In the commit message)
> We need to support `jj new`, and we need to add a `--branch` argument to `jj new`

It would be weird if `jj new --branch` created a branch on the parent of the newly created commit. And it's not possible if `new` created a merge. I think this is the same kind of problem that `jj git push -b` would have to push the parent if the current branch were advanced to `@`.

FWIW, when learning Git, I find it's confusing that `git checkout -b` is most likely the right command to create new branch, and `git branch` is NOT the command to start working on the created branch.

https://github.com/martinvonz/jj/pull/3129#issuecomment-1962774589
> We generally don't treat `@-` differently. Maybe that's all that bothers me about this proposal.

For some local commands like `jj log -r..<current-branch>`, it would be more useful if the current branch is pointing to `@`. For `git push`/`export` (and native push?), it's not. I have no idea how to model that in jj.

emesterhazy commented:
I think whatever we do, if anything, the model needs to be simple and intuitive. I don't think that we should, for example, have the branch set to `@` in jj but push `@-` to github, or have the branch set to `@` in jj and set the git ref for the branch to `@-`. I think that type of inconsistency will be cumbersome and confuse users.

I think there are probably three user "stories" for this feature:

1. I don't care about automatically advancing branches and will leave this feature disabled.
2. I am using jj in a **non-colocated** repo and am pushing multi-commit branches to a Git forge. When I am working on a branch, I want it to point to my last finished commit. When I am done working, I will have `@` set to an empty commit as the result of `jj commit` or `jj new`. I don't want to push this empty commit to my forge.
3. I am using jj in a git **colocated** repo and am pushing multi-commit branches to a Git forge. When I am working on a branch, I want it to point to my last finished commit. _If I use the `git` CLI, I want HEAD to be set to the branch instead of being detached so that `git` behaves normally._ When I am done working, I will have `@` set to an empty commit as the result of `jj commit` or `jj new`. I don't want to push this empty commit to my forge.

If you look at the original FR https://github.com/martinvonz/jj/issues/2338#issue-1928609981 I think it's asking for an easy way to set the "working" branch to `@-` automatically. I think that user probably falls into category 2.

Do you think there's a user story that's better served by advancing the branch to `@` instead of `@-`? What do you think it looks like?

martinvonz commented:
If there are unfinished commits on top of branches, I probably would want to see those changes when I run `jj log main..my-branch`, for example. That would require the branch to point to the latest commit, whether or not that commit is finished.

It sounds like this feature is only for Git interop. Your two no-op user stories were both about Git interop. So do you think we should simply put this in a `jj git commit` command? User who use that workflow a lot can create an alias for that command. That seems like a simple solution that doesn't involve trying to figure out a jj-native workflow for advancing branches.

BatmanAoD commented:
> So do you think we should simply put this in a `jj git commit` command? User who use that workflow a lot can create an alias for that command. That seems like a simple solution that doesn't involve trying to figure out a jj-native workflow for advancing branches.

I'm not sure how much sense that makes logically, but as one data point, I do pretty often mistype `jj commit` as `jj git commit`.

emesterhazy commented:
I think the question about whether the branch should move to `@` or `@-` is fundamentally about what a branch represents and which changes should be included. Right now it's completely up to the user because they need to manually move the branch. This is where my opinions come in, but I think that one of the most common uses of branches is to point to a line of "completed" work, and the more I think about that use case the more I'm convinced that advancing the branch to `@-` is the only way to make this feature ergonomic.

If we move the branch to `@`:
- In your `jj log main..my_branch` example you would see the empty or unfinished commit.
- If you run `jj new my_branch` it will create a second empty commit on top of the one my_branch points to.
- If you don't make any changes to the empty commit the branch points to and run `jj new` on a different revision, the empty commit will not be abandoned. Without a branch set to the empty commit, it would be abandoned.
- If you want to build on top of the last "completed" commit in my_branch you'll need to use `jj edit my_branch` instead of `jj new`
- If you are pushing a branch to a forge (git or otherwise), you'll need to manually move it behind the empty commit.

If we move the branch to `@-`:
- You can run `jj log (main..my_branch)::` to see any unfinished commits on top of your branch.
- You can run `jj new my_branch` to get a new empty commit on top of the branch. If you don't make any changes and run `jj new` on a different revision, the empty commit will be abandoned (this is the current default behavior).
- If you want to build on top of the last "completed" commit in my_branch, you can just run `jj new my_branch`. You don't need to switch to the `edit` command to avoid creating an extra empty commit.
- If you are pushing a branch to a forge, you don't need to move the branch before pushing to avoid including an empty commit.

jj doesn't distinguish between committed and uncommitted changes, but that doesn't mean that users don't treat the "working commit" at `@` differently than `@-`. Even jj itself treats the empty `@` commits created by `new` and `commit` differently since it will abandon them if you make no changes and run `jj new` on a different revision.

I think "branches point to finished work" for some definition of "finished work" that doesn't include empty commits created by `jj new` and `commit` is probably the most common use of branches. As long as we support a common and reasonable workflow I think we're in good shape. Users can always opt out of automatically advancing branches and have full control over the branch pointers.

Let's try to sort out the behavior and supported use cases first and then decide if this should be controlled by a config setting or flag on `jj new` and `jj commit`.

yuja commented:
> I think that one of the most common uses of branches is to point to a line of "completed" work, and the more I think about that use case the more I'm convinced that advancing the branch to `@-` is the only way to make this feature ergonomic.

Regarding "completed" work, I think `@-` is often unfinished (wip) commit under topic branch worflow, and for me, `@-` is not so different from `@` in that they are all draft changes. I would expect `jj log -r main@origin..my_branch` shows the stack of draft changes including `@` if my_branch auto-advances.

> If we move the branch to `@`:
[...]
> * If you don't make any changes to the empty commit the branch points to and run `jj new` on a different revision, the empty commit will not be abandoned. Without a branch set to the empty commit, it would be abandoned.

I made this change to support a workflow where named branch is set at the root of the stack #1854, not at the head (which would require auto-advancing.) If we decide to set the active branch to `@`, it might be better to revert the change (or at least make it configurable.)

> * If you are pushing a branch to a forge (git or otherwise), you'll need to manually move it behind the empty commit.

Yes, this is the problem. It's unclear whether the `@` commit should be pushed if it's not empty.

martinvonz commented:
> If we decide to set the active branch to `@`, it might be better to revert the change (or at least make it configurable.)

+1

> * If you want to build on top of the last "completed" commit in my_branch, you can just run `jj new my_branch`. You don't need to switch to the `edit` command to avoid creating an extra empty commit.

OTOH, you will then start a new commit when you might have wanted to continue working on the unfinished commit you already had on top of `my_branch`.

> > * If you are pushing a branch to a forge (git or otherwise), you'll need to manually move it behind the empty commit.
> 
> Yes, this is the problem. It's unclear whether the `@` commit should be pushed if it's not empty.

I think `jj git push` should not push it. That would better match Git's behavior.

I don't know what I would expect non-Git forges to do. FWIW, when exporting to our internal forge at Google, we do export the `@` commit too, but it's too early to say yet if users will like that behavior.

I'm still not sure which is the right choice. I don't want to block progress on this. As you long as you don't change the default behavior too much, we can just get something in so people can experiment with it.

emesterhazy commented:
I'm starting to think there's no perfect solution here since there are many possible workflows. I still think that `@-` is the "lesser of two evils", but that's as much as it is since you can conceive of a useable workflow where branches are advanced to `@` as well.

So, I would just like to acknowledge that there are lots of different opinions here, and we may not be able to come up with a solution that's simple and useful for everyone.

@martinvonz @yuja if you're interested in submitting something experimental so that users can try it without needing to build `jj` from this branch, here's what I'd suggest:

1. I will finish the code for `jj new` and add more tests.
2. Can we name the config option `experimental-advance-branches` to make it clear that this feature may change or be removed?
3. Whether or not to advance to `@` or `@-` doesn't actually make much difference in the code. I can add another config option to control that behavior.
4. If we add the `@` config option, I can look into disabling the behavior from https://github.com/martinvonz/jj/issues/1854

martinvonz commented:
SGTM. I think we can ignore items 3 and 4 until people have experimented with the feature for a while. I don't want you to waste time implementing item 3 if it seems like everyone is happy with what you already have in this PR.

Thanks for all the time you've spent thinking and coding on this.

yuja commented:
> 1. I will finish the code for `jj new` and add more tests.
> 2. Can we name the config option `experimental-advance-branches` to make it clear that this feature may change or be removed?
> 3. Whether or not to advance to `@` or `@-` doesn't actually make much difference in the code. I can add another config option to control that behavior.

If we choose the `@` model, we won't need the `advance-branches` config. Having a branch on `@` means the branch is current/active, and it will advance. In other words, `HEAD = <branch>` (not detached) in git is encoded as `@ = <branch>, @- = <branch>@git` in jj. Anyway, I agree that trying 3 isn't needed. IMHO, it would be quite different from the `@-` design.

I'm skeptical about `jj new --branch` because it wouldn't make sense if `experimental-advance-branches` is off.

BatmanAoD commented:
Tried this out very briefly just now, and it looks like it does exactly what I was expecting. I'll be using it this week and see if I have further feedback. Thanks very much for implementing!

BatmanAoD commented:
This is excellent. It's exactly what I want.

emesterhazy commented:
Thanks for the feedback. I'll add support for `jj new` as well. Just to be clear, this means that if `jj new` is run on revision `xyz` and one of the parents of `xyz` has an advanceable branch pointing to it, then the branch will be advanced to `xyz`. It is analogous to how `jj commit` works, except that you can specify a revision.

> I'm skeptical about jj new --branch because it wouldn't make sense if experimental-advance-branches is off.

I'm not planning to add any flags to `jj new` or other commands. I think in an earlier version I proposed adding a flag to disambiguate in scenarios where multiple branches might advance, but I think we should just make the user manually move the branch with `jj branch set` in those cases. So now my plan is to have this controlled exclusively by the config.

emesterhazy commented:
@martinvonz

Hi everyone, sorry for the delay. This is ready for review now. A couple of notes:

1. The config setting is called `experimental-advance-branches`. If we decide to stabilize this we can remove the experimental prefix.
2. I added support for `jj new`. It will only advance branches if a single revision is targeted and neither `--after` or `--before` are used. I think this is a reasonable limitation to keep the behavior simple. In the common case where commits are usually appended, this limitation doesn't matter.
3. This PR does not include the change to not detach the Git head for colocated repos. I will mail a separate PR after this one is submitted. That PR will also include updates to the documentation and changelog.
4. As I mentioned in a [comment](https://github.com/martinvonz/jj/pull/3129#issuecomment-1984444454) earlier, I changed the semantics so that multiple branches can advance at the same time. This ends up being simpler for the user since they don't need to worry about resolving ambiguity.
5. I updated the PR description based on the new updated semantics.

yuja commented on cli/tests/test_advance_branches.rs:
```diff
@@ -40,9 +42,37 @@ fn set_advance_branches_overrides(
+        }
+    )*
+    }
+}
```
I think this will make harder for us to maintain the tests.

Since `jj commit` is deprecated (and is practically the same as `jj new`), I don't think we need to implement the advance-branches feature in `jj commit`. That will simplify the testing scenario.

BatmanAoD replied:
I've been using `commit` almost exclusively. If `new` and `commit` diverge in such a way that `commit` will, without warning, leave my branch behind, that will be very frustrating.

yuja replied:
Oops, maybe I confused it with the `checkout` deprecation. I also use `jj ci` often.

Anyway, I prefer tests without too much meta programming. If needed, `test_case()` macro could be used, but parameterized tests generally don't work well with insta.

emesterhazy replied:
Thanks, I didn't know we had `test_case`. I'm going to replace the custom macro with that. insta is cooperating as long as we wrap the expectations in allow_duplicates. I think it's worth parameterizing here since the test expectations are the same.

yuja commented on cli/src/cli_util.rs:
```diff
@@ -1312,6 +1325,51 @@ Then run `jj squash` to move the resolution into the conflicted commit."#,
+    /// Returns an empty `std::Vec` if no branches are eligible to advance.
+    pub fn get_advanceable_branches<'a>(
+        &self,
+        repo: &impl Repo,
```
`repo` is known to the `WorkspaceCommandHelper`. Passing it as argument will obfuscate the scope of the `WorkspaceCommandHelper` methods.

emesterhazy replied:
Thanks... not sure why I did this. I think it was a  carryover from an earlier draft and I didn't realize it wasn't necessary anymore.

yuja commented on cli/src/cli_util.rs:
```diff
@@ -1312,6 +1325,51 @@ Then run `jj squash` to move the resolution into the conflicted commit."#,
+        let advance_branches = self.settings.advance_branches();
+        let overrides = self.settings.advance_branches_overrides();
+        let allow_branch = |branch: &str| {
+            let overridden = overrides.iter().any(|x| x.eq(branch));
```
nit `x == branch`

emesterhazy replied:
Done.

yuja commented on cli/src/cli_util.rs:
```diff
@@ -1312,6 +1325,51 @@ Then run `jj squash` to move the resolution into the conflicted commit."#,
+        let allow_branch = |branch: &str| {
+            let overridden = overrides.iter().any(|x| x.eq(branch));
+            (advance_branches && !overridden) || (!advance_branches && overridden)
+        };
```
Maybe better to add positive/negative patterns instead?
```toml
advance-branches-includes = ["glob:*"]
advance-branches-excludes = ["exact:main"]
```

Suppose the user configured `experimental-advance-branches.enabled = true` globally, he might add `"main"` to the repo-level `overrides` list. If he changes mind later and turns it off globally, the "main" branch will start advancing. That's scary.

emesterhazy replied:
Thanks for this suggestion, I agree. I updated the PR so that an example config looks like this:

```
[experimental-advance-branches]
enabled-branches = ["glob:*"]
disabled-branches = ["main"]
```

yuja commented on cli/src/cli_util.rs:
```diff
@@ -1312,6 +1325,51 @@ Then run `jj squash` to move the resolution into the conflicted commit."#,
+        }
+
+        let mut advanceable_branches = Vec::new();
+        for from_commit in from.into_iter() {
```
nit: `.into_iter()` can be omitted

emesterhazy replied:
Thanks, done.

yuja commented on cli/src/commands/commit.rs:
```diff
@@ -54,6 +54,8 @@ pub(crate) fn cmd_commit(
         .ok_or_else(|| user_error("This command requires a working copy"))?;
     let commit = workspace_command.repo().store().get_commit(commit_id)?;
+    let advanceable_branches = workspace_command
+        .get_advanceable_branches(&**workspace_command.repo(), commit.parent_ids());
```
Git user might prioritize the first parent, but I'm not sure. It's unclear whether the branches of a merge commit should advance.

BatmanAoD replied:
The Git CLI represents a merge as a "directed" operation, merging one branch "into" another. The branch beingmmerged "into" is advanced; the other is not. The closest analogue here, I think, would be to advance any branches in `@-` but not in other merge targets. That introduces a dependency on the current working commit state, though, which I think would be weird when making a merge with `new`. Arguably there should be a convenience command that's closer to `git merge` anyway, though; in particular, there's currently no easy way to express "create a new commit only if there are commits I don't already have on my branch", which is the default `git` behavior (and very convenient, especially since it makes `git pull origin main` idempotent).

yuja replied:
> advance any branches in `@-` but not in other merge targets.

My concern here is what to do if the current working-copy commit is a merge. `@-` points to multiple parents. Perhaps, git will move the first parent branch (because `HEAD` can't point to multiple revisions)?

emesterhazy replied:
I thought of an alternative. What if we don't advance any branches if there are multiple candidates and instead print a warning for the user indicating that no branches were advanced because multiple candidates were found.

BatmanAoD replied:
Thinking about this again: the typical workflow and desired behavior is something like:

```bash
# "merge from" the branch you want (which will often be `main`); starting from a clean working copy:
> jj new @- <other-branch>
# Now `@` is a merge commit; in Git, the branch that previously pointed to `@-` would now point to `@`
> jj commit -m "Merge from <other-branch> into <current-branch>"
```

If `jj new` with multiple branches doesn't advance at least one of those branches, then you end up with ambiguity. But when `jj new @- <other>` is executed, it's obvious which branch should be advanced (the one pointing to `@-`).

So I wonder if `new @- <others>` should be treated as a special case by advancing the branch from `@-`.

(Again, though, I think it also highlights the usefulness of a separate `merge` command that would do both of the above steps, creating the merge commit _and_ moving the working copy to a new, non-merge commit.)

emesterhazy replied:
I like this suggestion, but it's worth noting that it introduces a slight discrepancy into the model. Currently `jj new` and `jj commit` create a new commit on top of the target and advance the branch pointing to the target's parent to the target, not to the new commit.

Your suggestion is to make merges a special case and advance the branch from the first target to the new commit if a merge commit is created. That seems reasonable to me and better than doing nothing and making the user resolve it. The reason I think this is ok is because a merge commit is usually "finished" as soon as it's created and can be pushed to a remote. A new empty commit isn't usually "finished" and can't or won't be pushed to a remote.

It would basically work like this:
```
jj log
◉  branch2
│ @  branch1
├─╯
◉
jj new -m "merge commit" branch1 branch2
jj log
@    branch1 merge commit
├─╮
│ ◉
◉ │  branch2
├─╯
◉
```

If Yuya agrees I think we should go with this design and I can update the PR.

yuja replied:
> Your suggestion is to make merges a special case and advance the branch from the first target to the new commit if a merge commit is created. That seems reasonable to me and better than doing nothing and making the user resolve it. The reason I think this is ok is because a merge commit is usually "finished" as soon as it's created and can be pushed to a remote. A new empty commit isn't usually "finished" and can't or won't be pushed to a remote.
> 
> It would basically work like this:
> 
> ```
> jj log
> ◉  branch2
> │ @  branch1
> ├─╯
> ◉
> jj new -m "merge commit" branch1 branch2
> jj log
> @    branch1 merge commit
> ├─╮
> │ ◉
> ◉ │  branch2
> ├─╯
> ◉
> ```

Suppose the user did `jj new` to finish the merge, doesn't it move the branch2 to the merge commit?
```
jj new @
jj log
@
◉    branch1 branch2 merge commit
├─╮
│ ◉
◉ │
├─╯
◉
```

That was my original concern, and I would argue against adding another special case. If we need to support Git-like merge behavior, I think it's better to add a separate merge command something like `jj git-like-merge <arg>` == `jj new @ <arg>; jj new; jj branch set -r@- ..`.

BatmanAoD replied:
This feature probably shouldn't _ever_ advance a branch from `@-` to `@` if `@` already has a branch pointing to it. I don't think that's really a special case. 

Side-note: IMO there's nothing especially "git-like" about distinguishing between "merge from" and "merge into".

yuja replied:
> This feature probably shouldn't _ever_ advance a branch from `@-` to `@` if `@` already has a branch pointing to it. I don't think that's really a special case.

I meant this part is the special case: "advance the branches from the first target *to the new commit* if a merge commit is created." If the new commit isn't a merge, `jj new` will advance the branch *to the parent* of the new commit.

> Side-note: IMO there's nothing especially "git-like" about distinguishing between "merge from" and "merge into".

Git has a notion of the current branch (or symbolic `HEAD`.) I think that's the main difference. `jj new` creates a merge of arbitrarily-chosen commits, whereas `git merge` will merge the current commit with the others.

emesterhazy replied:
> This feature probably shouldn't ever advance a branch from @- to @ if @ already has a branch pointing to it. I don't think that's really a special case.

Currently it does, but maybe you're right and it shouldn't.

> I meant this part is the special case: "advance the branches from the first target to the new commit if a merge commit is created." If the new commit isn't a merge, jj new will advance the branch to the parent of the new commit.

Yep, it's a special case. I don't think that's a big deal though. It should be obvious that a merge commit and a new empty commit are pretty different. I don't think it will be surprising to users that the feature behaves this way after they've seen it once, and I think it makes the feature more useful, even if it's slightly more complex as a result.

BatmanAoD replied:
@yuja When you said "I would argue against another special case", I thought you meant that there would be two special cases involved in handling merges: one when advancing the branch during the merge itself, and one when advancing past the merge. I just meant that with the additional rule of "don't auto-advance to something that's already a different branch", there's only one special case here, not two; but I do agree that the idea we're discussing is a special case.

@emesterhazy Yes, I should have clarified that I am suggesting "don't auto-advance to an existing branch" as a new change.

yuja replied:
> It should be obvious that a merge commit and a new empty commit are pretty different.

I originally thought that way when I switched from `git`/`hg` to `jj`, but I changed mind. So I think the proposed behavior would be surprising to people who learned `jj` a lot.

BatmanAoD replied:
I have to say, I still don't understand the "jj way", then. Git is effectively the only usable backend outside of Google, and to use git, you need one of:

1. A fork for everything, so you can commit directly to `main`
2. A habit of using `push -a` and later cleaning up after yourself
3. Branches

With branches, there's absolutely a difference between merge commits and non-merge commits. So how *should* I be thinking about this?

emesterhazy replied:
My current opinion is that this feature already operates on the margins of what might be considered the "native" `jj` workflow. Given that, I think it's ok if it's a little impure or surprising to `jj` experts as long as it is reasonable, consistent, and useful.

This feature is experimental, so we could submit it as-is and request feedback from potential users in #2338. I think we want to resolve these questions:

1. Should we leave the branches as-is if there are multiple branches eligible to advance?
2. Should we only advance a branch if there isn't already a branch pointing to the target commit?
3. Should we add a special case for new commits with multiple parents (aka merges) at moves a branch to the merge commit?

I think my vote for all three of these is yes. Maybe only number 3 is controversial though? If there's agreement on 1 and 2 then I'd like to implement those before I finish https://github.com/martinvonz/jj/pull/3405 and add information about this feature to the changelog.

yuja replied:
> 1. Should we leave the branches as-is if there are multiple branches eligible to advance?

(clarification: multiple parents, not multiple named branch refs in a single parent?)

> 2. Should we only advance a branch if there isn't already a branch pointing to the target commit?
> 3. Should we add a special case for new commits with multiple parents (aka merges) at moves a branch to the merge commit?

My vote: 1. Maybe?, 2. Maybe?, 3. No.

I hope there will be as few special cases as possible, and to me, 3 is the weirdest one.

However, I'm not the target audience of this feature. I don't use named branches locally, and named branches are created only by `jj git push -c`.

BatmanAoD replied:
@emesterhazy Could you please implement #2, at least? Currently, as far as I can tell, there's no way to pick which branch to advance, so I have to either explicitly push one of them backwards after `jj new`, or, I suppose, temporarily muck with my config.

Another idea is to not advance anything when the user runs `jj new <branchname>`. This would be a more explicit version of not advancing branches when there's already a branch at the destination.

emesterhazy replied:
I think we should merge this PR now. I've been sitting on it for a while since I'm not sure if this is the best long-term solution for branches. There are [other discussions](https://github.com/martinvonz/jj/issues/3505) and proposals for how branching/bookmarking/topics/aliases etc should work. But, since this is meant to be experimental, I think we can submit it and rip it out later if we decide on a different direction.

---

RE: Implementing number 2, doing this the way I wanted ended up being a little tricky for `commit` since any existing branches are moved to the rewritten `@` commit after the branches are advanced for `advance-branches`.

I have a WIP PR, but I haven't completely gotten it working yet: https://github.com/martinvonz/jj/pull/3548

yuja commented on cli/tests/test_advance_branches.rs:
```diff
@@ -0,0 +1,236 @@
+            &format!("[{}]", override_string),
+        ],
+    )
+}
```
nit: you can use `test_env.add_config()` (to enable it globally) or `--config-toml` (temporarily).

emesterhazy replied:
Thanks, I wasn't aware of these options. `add_configs` looks like it will work, but I'm going to make the change in the second config. If it make it in the first config then all of commits will have new change ids and the snapshots will break. This will cause a lot of annoying merge conflicts in the second commit when the tests are parameterized.

yuja reviewed (commented):
This generally looks good to me, thanks.

emesterhazy commented:
I've made changes and resolved most of the conversations based on your comments. Please take another look. Thanks!

yuja commented on cli/src/cli_util.rs:
```diff
@@ -1362,6 +1375,96 @@ Then run `jj squash` to move the resolution into the conflicted commit."#,
+    ) -> Result<Vec<StringPattern>, CommandError> {
+        let setting = format!("experimental-advance-branches.{setting_key}");
+        let err_msg = format!("Error parsing pattern for {setting}");
+        match self.settings.config().get_array(&setting) {
```
nit: error handling can be simplified by `config().get::<Vec<String>>(key).optional()?`

emesterhazy replied:
Looks good to me, thanks!

yuja commented on cli/src/cli_util.rs:
```diff
@@ -1362,6 +1375,96 @@ Then run `jj squash` to move the resolution into the conflicted commit."#,
+        from: impl IntoIterator<Item = &'a CommitId>,
+    ) -> Result<Vec<AdvanceableBranch>, CommandError> {
+        let enabled_branches = self.advance_branches_enabled_branches()?;
+        let disabled_branches = self.advance_branches_disabled_branches()?;
```
nit: might be better to add a separate `AdvanceBranchesSettings`/`Patterns` type and move parsing functions there. `WorkspaceCommandHelper` tends to get messier.

```rust
struct AdvanceBranchesSettings {
    enabled_branches,
    disabled_branches,
}

impl AdvancedBranchesSettings {
    fn from_config(config) -> Result<Self, CommandError>
}
```

emesterhazy replied:
Yeah, good idea. I didn't really like how this cluttered up the workspace helper either.

yuja commented on cli/src/cli_util.rs:
```diff
@@ -1362,6 +1375,96 @@ Then run `jj squash` to move the resolution into the conflicted commit."#,
+                if enabled.matches(branch) {
+                    return true;
+                }
+            }
```
nit: `.iter().any()` can be used

emesterhazy replied:
Thanks, that's cleaner.

yuja commented on cli/src/cli_util.rs:
```diff
@@ -1448,6 +1551,34 @@ impl WorkspaceCommandTransaction<'_> {
+                .simplify(),
+            );
+            self.mut_repo()
+                .set_local_branch_target(&branch.name, new_target);
```
Perhaps, `.merge_local_branch()` can be used.

```rust
self.mut_repo().merge_local_branch(
    &branch.name,
    &RefTarget::normal(branch.old_commit_id.clone()),
    &RefTarget::normal(move_to.clone()),
);
```

(I thought I added this comment in the previous round, but apparently not.)

emesterhazy replied:
Thanks, this works perfectly.

yuja commented on cli/src/commands/commit.rs:
```diff
@@ -119,6 +120,11 @@ new working-copy commit.
             )
             .write()?;
+
+        if !advanceable_branches.is_empty() {
```
nit: emptiness check is redundant (though it's harmless.)

emesterhazy replied:
Yeah... I added it intentionally to make it clearer to the reader that nothing happens if there's no branches to advance. Although I guess that might be obvious anyways. I'll delete and add a short comment.

yuja commented on lib/src/settings.rs:
```diff
@@ -288,6 +288,12 @@ impl UserSettings {
+        self.config
+            .get_bool("experimental-advance-branches.enabled")
+            .unwrap_or(false)
+    }
```
nit: delete

emesterhazy replied:
Oops, I thought I removed this. Sorry!

yuja commented on cli/src/commands/new.rs:
```diff
@@ -206,6 +209,27 @@ Please use `jj new 'all:x|y'` instead of `jj new --allow-large-revsets x y`.",
+        return ws.get_advanceable_branches(target_commits[0].parent_ids());
+    } else {
+        Ok(Vec::new())
+    }
```
nit: maybe simpler to just inline this function? The caller has to know that `target_commits` is singular if `advanceable_branches` is not empty.

Alternatively, the assumption can be encoded as `Option<(Vec<AdvanceableBranch>, CommitId)>`, but I'm not sure if that's better.

emesterhazy replied:
Personally I'm not too worried about this since there's a comment explaining and the function is easy to read. I just wanted to pull the logic out of the main function to make it more readable. In general I think many of the command functions are a bit too long and would be easier to read if we pulled some of the details into functions.

If you want I'm fine inlining this, but I also think it's ok as-is.

yuja commented on cli/src/cli_util.rs:
nit: stale comment in the commit description
> WIP: use patterns instead of overrides list to control advance-branches

emesterhazy replied:
Thanks. That should have been removed when I squashed the WIP commit, but I guess I lost track of it.

yuja commented on cli/tests/test_advance_branches.rs:
```diff
@@ -1,116 +1,133 @@
-        enable_advance_branches_for_patterns(test_env, cwd, &[""]);
+        test_env.add_config(
+            r#"[experimental-advance-branches]
+        enabled-branches = [""]
```
nit: an empty array `[]` also works

emesterhazy replied:
Yep, done.

yuja commented on cli/src/commands/new.rs:
```diff
@@ -199,6 +201,26 @@ Please use `jj new 'all:x|y'` instead of `jj new --allow-large-revsets x y`.",
+    let should_advance_branches =
+        target_commits.len() == 1 && !args.insert_before && !args.insert_after;
+    if should_advance_branches {
+        return ws.get_advanceable_branches(target_commits[0].parent_ids());
```
nit: redundant `return`

emesterhazy replied:
Good point.

yuja reviewed (approved):
thank you for working this feature!

ilyagr commented:
For people interested in this feature, take a look at the discussion at https://github.com/jj-vcs/jj/discussions/3549 as well.
